### PR TITLE
Corrected value returned in the id field of the RPC response

### PR DESCRIPTION
**Pull request by maciejzaleski**
In case of the single RPC request the RPC response used hard-coded `id=1`. This changes streamlines the response processing so that it behaves the same way for batch and single request.

The data-type for the id field in the response will match the data-type used in the request.

### CODE REVIEW

**Comment by dangershony** [comment]
To try to fix this issue, but it seems this PR does not fix the issue (yet)
https://github.com/stratisproject/StratisBitcoinFullNode/issues/3921

**Comment by fenix2222** [comment]
There are 5 failing unit tests and fairly heavy code refactor that needs to be thoroughly tested. I am not sure if current code refactor has no issues as it needs better code coverage.

I have created a much simpler fix that we can put in place for now and it passes all the tests including manual tests I have performed: https://github.com/stratisproject/StratisBitcoinFullNode/pull/3927

**Comment by maciejzaleski** [comment]
Test `WriteResponseBodyAsyncWritesContextToResponseBody` relies on memory leak to function. I don't think this is a good idea.

**Comment by maciejzaleski** [comment]
Fixed failing RPC tests and re run locally to confirm that all RPC tests are passing.

In addition I have manually tested the RPC calls:
```
Calling non-existent RPC method
Request: {"method": "getblockcound", "id": "abc"}
Old response: {"result": null, "error": { "code": -32601, "message": "Method not found"}}
New response: {"result":null,"error":{"code":-32601,"message":"Method not found"},"id":"abc"}

Using boolen as id argument and obtaining result as primitive type
Request: {"method": "getblockcount", "id": true}
Old response: {"result":20494,"id":1,"error":null}
New response: {"result":17899,"error":null,"id":true}

Calling non-existing RPC method as a single item batch
Request: [{"method": "getblockcound", "id": "abc"}]
Old response: [{"result":null,"error":{"code":-32601,"message":"Method not found"},"id":"abc"}]
New response: [{"result":null,"error":{"code":-32601,"message":"Method not found"},"id":"abc"}]

Calling three methods where one is non-existing RPC method, one has invalid data-type for its input parameter and one has been called correctlyparameter 
Request: [{"method": "getblockcount1", "id": "abc"},{"method": "getblockcount", "id": "xyz"},{"method": "getrawtransaction", "id": "qqq", "params": [11]}]
Old response: [{"result":null,"error":{"code":-32601,"message":"Method not found"},"id":"abc"},{"result":20495,"id":"xyz","error":null},{"result":null,"error":{"code":-1,"message":"Argument error: trxid"},"id":"qqq"}]
New response: [{"result":null,"error":{"code":-32601,"message":"Method not found"},"id":"abc"},{"result":17899,"error":null,"id":"xyz"},{"result":null,"error":{"code":-1,"message":"Argument error: trxid"},"id":"qqq"}]

Calling multiple methods with incorrect parameters; the response is returned for each item in the batch
Request: [{"method": "getblockhash", "id": "abc", "params": [100]},{"method": "getblockhash", "id": "xyz", "params": []},{"method": "getblockhash", "id": "www", "params": 100},{"method": "getblockcount1", "id": "xyz"},{"method": "getrawtransaction", "id": "qqq", "params": ["47511f10dc8106299c710cc949a6616ef57b84921160766c15516c27b0bdaf3c"]}]
Old response: [{"result":"46c54bcf2bc05cdc153bf6e7b1fdbfbda53a43c5f273ce578e5808d116b47316","id":"abc","error":null},{"result":"0000066e91e46e5a264d42c89e1204963b2ee6be230b443e9159020539d972af","id":"xyz","error":null},{"result":null,"error":{"code":-32603,"message":"Internal error"},"id":"www"},{"result":null,"error":{"code":-32601,"message":"Method not found"},"id":"xyz"},{"result":"01000000a526cf5b01411bb2f69ba1ca959c484d3b40803dcbfcc95c368f300adfb26bad1a69c3edfb000000006b483045022100974695915c65c2f6cf1a9ba085349585860287afe34ee5567b0df47a0358f12a022029be7bd4923b6c3e20d6aba0ff3b48ad46a4861668c959403451ea1d8d0e3bdf0121038a3b15d4a148ccec9b57351759ee1da9b7eb533667c2e58846d4e75908689c6dfeffffff0240420f00000000001976a91467239655458d705480009aff59ee35f67240f4e588acb01ac817000000001976a91497ed7bbf2bb30e4c8c654cc14dcd76cd4714de0488ac00000000","id":"qqq","error":null}]
New response: [{"result":"46c54bcf2bc05cdc153bf6e7b1fdbfbda53a43c5f273ce578e5808d116b47316","error":null,"id":"abc"},{"result":"0000066e91e46e5a264d42c89e1204963b2ee6be230b443e9159020539d972af","error":null,"id":"xyz"},{"result":null,"error":{"code":-32603,"message":"Internal error"},"id":"www"},{"result":null,"error":{"code":-32601,"message":"Method not found"},"id":"xyz"},{"result":"01000000a526cf5b01411bb2f69ba1ca959c484d3b40803dcbfcc95c368f300adfb26bad1a69c3edfb000000006b483045022100974695915c65c2f6cf1a9ba085349585860287afe34ee5567b0df47a0358f12a022029be7bd4923b6c3e20d6aba0ff3b48ad46a4861668c959403451ea1d8d0e3bdf0121038a3b15d4a148ccec9b57351759ee1da9b7eb533667c2e58846d4e75908689c6dfeffffff0240420f00000000001976a91467239655458d705480009aff59ee35f67240f4e588acb01ac817000000001976a91497ed7bbf2bb30e4c8c654cc14dcd76cd4714de0488ac00000000","error":null,"id":"qqq"}]

Calling multiple methods where one returns a complex data type
Request: [{"method": "getblockcount", "id": "abc"},{"method": "getblockcount", "id": "xyz"},{"method": "gettransaction", "id": "qqq", "params": ["47511f10dc8106299c710cc949a6616ef57b84921160766c15516c27b0bdaf3c"]}]
Old response 1: [{"result":20503,"id":"abc","error":null},{"result":20503,"id":"xyz","error":null},{"result":null,"error":{"code":-5,"message":"Invalid or non-wallet transaction id."},"id":"qqq"}]
Old response 2: [{"result":20503,"id":"abc","error":null},{"result":20503,"id":"xyz","error":null},{"result":{"amount":10000.0,"confirmations":13503,"blockhash":"d1f9e7c5981c349487261dfc3b81525906da34f5641bfab6d965a320b526ae2c","blockindex":4,"blocktime":1559116096,"txid":"5e1290769ee91dd2f58349b83088998172dc5924d56d6e56eb6ca8127f6ecdf7","time":1558788492,"timereceived":1558788492,"details":[{"address":"SgKKjTJBAEk1WXMYTeR5XHCF1cVsAFR8D8","category":"receive","amount":10000.0,"vout":0}],"hex":"010000008c39e95c01f27f83e13f0789fabece01a43805d6a7f5f616fe6cf846b17810199076b9ee6b0100000048473044022075cc7b81e7ce762d6f5f33fe2d9ef0891ec785b91bb4b11ef57ec62b56977d4702206d01e60498df4e81ec3b4d4484f21adfccae8c60af40d437e70ffe5c48eb36a201ffffffff020010a5d4e80000001976a914d0aa6a7f56bca20626dc925937ab66a23bfc31e588acccb9a75c060100001976a9145c6ce150cc4d23188971f6e456414bd2b6fbf23188ac00000000"},"id":"qqq","error":null}]
New response 1: [{"result":17899,"error":null,"id":"abc"},{"result":17899,"error":null,"id":"xyz"},{"result":null,"error":{"code":-5,"message":"Invalid or non-wallet transaction id."},"id":"qqq"}]
New response 2: [{"result":17899,"error":null,"id":"abc"},{"result":17899,"error":null,"id":"xyz"},{"result":{"amount":0.01,"confirmations":17838,"blockhash":"08ec3da3390e29b025afb9950892de3cc01770dc099644ff95bc341df4cf1e9d","blockindex":9,"blocktime":1540302502,"txid":"47511f10dc8106299c710cc949a6616ef57b84921160766c15516c27b0bdaf3c","time":1540302501,"timereceived":1540302501,"details":[{"address":"SWhMGFAbH5tPAyVE8F8TpgytXTC9m3PcTk","category":"receive","amount":0.01,"vout":0}],"hex":"01000000a526cf5b01411bb2f69ba1ca959c484d3b40803dcbfcc95c368f300adfb26bad1a69c3edfb000000006b483045022100974695915c65c2f6cf1a9ba085349585860287afe34ee5567b0df47a0358f12a022029be7bd4923b6c3e20d6aba0ff3b48ad46a4861668c959403451ea1d8d0e3bdf0121038a3b15d4a148ccec9b57351759ee1da9b7eb533667c2e58846d4e75908689c6dfeffffff0240420f00000000001976a91467239655458d705480009aff59ee35f67240f4e588acb01ac817000000001976a91497ed7bbf2bb30e4c8c654cc14dcd76cd4714de0488ac00000000"},"error":null,"id":"qqq"}]

```

**Comment by fenix2222** [comment]
Looks much better now.

**Comment by dangershony** [comment]
There are a lot of changes in code, and I did not go in depth but not sure what this is trying to fix.
You already found the issue in electrumx and its not related to RPC.

Instead of adhering to standards  best to is adhere to how bitcoin core works are many services use those RPC endpoints

**Comment by maciejzaleski** [comment]
As per my comment above as far as I can tell all RPC calls behave the same except for one case where a defect have been fixed.

It would be most welcome if you could write a test which would highlight issue with the new implementation. I cannot think of any other way in which the RPC method could be called.

I have not verified that the RPC could be invoked from Swagger and I suspect that this might have relied on the hack from @NicolasDorier in the RPCRouteHandler which I have modified as I couldn't make sense out of it.

Here is a summary of changes
- Unified execution path for single and batched RPC calls so that they follow the same flow
- Ensured that the same response headers are used regardless of the execution path (i.e. single, batch, error)
- Ensured correct logic for the `id` parameter in RPC response (i.e. support for multiple data-types, if no `id` is passed in request the response won't contain it either)
- Ensured that the memory streams are closed
- Used adequate methods for transposing POCO into Json instead of re-writing the data between memory streams
- Used JObject/JToken objects to compare Json instead of comparing Json strings

File: src/Stratis.Bitcoin.Features.RPC/RPCRouteHandler.cs [hunk 1]
```diff
@@ -40,21 +40,21 @@ public async Task RouteAsync(RouteContext context)
         {
             Guard.NotNull(context, nameof(context));
 
-            var ms = new MemoryStream();
-            await context.HttpContext.Request.Body.CopyToAsync(ms);
-            context.HttpContext.Request.Body = ms;
-            ms.Position = 0;
-            JObject req = JObject.Load(new JsonTextReader(new StreamReader(ms)));
-            ms.Position = 0;
-            string method = (string)req["method"];
```
Comment (NicolasDorier) [inline src/Stratis.Bitcoin.Features.RPC/RPCRouteHandler.cs]: This code I wrote made sure that the `Body` was still readable after `RouteAsync`, by default body are not readable again. (Except if calling `HttpRequest.EnabledRewind` and rewinding it, but I don't think it existed then)

Those if things don't break even without it, I guess this is good.

Comment (maciejzaleski) [reply src/Stratis.Bitcoin.Features.RPC/RPCRouteHandler.cs]: I've just confirmed that there is no issue when calling the RPC methods through swagger.

Comment (maciejzaleski) [reply src/Stratis.Bitcoin.Features.RPC/RPCRouteHandler.cs]: Thanks for the explanation, I though there might have been a valid reason for this construction but looks like we can simplify the code now.

**Pull request merged**